scripts/data/src/org/akvo/gae/remoteapi/FixOrphanedSubmissions.java

```diff
@@ -0,0 +1,156 @@
+        }
+
+
+        Set<Long> dataPointsFound = findDataPointsIdsWithExistingRegistrationSurveyInstance(ds, surveyIds);
+        dataPointIds.removeAll(dataPointsFound);
+
+        ArrayList<Long> dataPointIdsToDelete = new ArrayList<>(dataPointIds);
+        List<Entity> instances = getSurveyInstancesToDelete(ds, dataPointIdsToDelete);
+
+        final StringBuilder sb = new StringBuilder();
+        if (instances != null) {
+            for (Entity instance : instances) {
+                Key key = instance.getKey();
+                surveyInstances.add(key);
+                sb.append(
+                        String.format("%s,%s,%s", instance.getProperty("surveyId"), instance.getProperty("surveyedLocaleIdentifier"), key.getId()))
+                        .append("\n");
+            }
+            FileUtils.write(f, sb.toString(), true);
+        }
+        if (surveyInstances.size() > 0) {
+            System.out.printf("Found a total of %d orphaned instances\n", surveyInstances.size());
+            if (doIt) {
+                batchDelete(ds, surveyInstances);
+                batchDelete(ds, dataPointIdsToDelete, "SurveyedLocale");
+            }
+        } else {
+            System.out.println("No orphaned SurveyInstances found");
+        }
+    }
+
+    /**
+     * Returns a list of datapoint ids
+     *
+     * We fetch all survey instances for given form (surveyId) and use their surveyedLocaleId to create the list of
+     * datapoints which we know have an existing SurveyInstance for them
+     */
+    private Set<Long> findDataPointsIdsWithExistingRegistrationSurveyInstance(DatastoreService ds, Set<Long> surveyIds) {
+        Set<Long> dataPointIds = new HashSet<>();
+        Set<Entity> surveyInstances = new HashSet<>();
+        for (Long surveyId : surveyIds) {
+            Query.Filter f = new Query.FilterPredicate("surveyId", Query.FilterOperator.EQUAL, surveyId);
+            Query.Filter f2 = new Query.FilterPredicate("surveyedLocaleId", Query.FilterOperator.NOT_EQUAL, null);
+            Query instanceQuery = new Query("SurveyInstance").setFilter(Query.CompositeFilterOperator.and(f, f2));
+            List<Entity> entities = ds.prepare(instanceQuery).asList(FetchOptions.Builder.withChunkSize(1000));
+            if (entities != null) {
+                surveyInstances.addAll(entities);
+            }
+        }
+        for (Entity e: surveyInstances) {
+            Long surveyedLocaleId = (Long) e.getProperty("surveyedLocaleId");
+            if (surveyedLocaleId != null) {
+                dataPointIds.add(surveyedLocaleId);
+            }
+        }
+        return dataPointIds;
+    }
+
+    private List<Entity> getSurveyInstancesToDelete(DatastoreService ds, List<Long> surveyedLocaleIds) {
+        if (!surveyedLocaleIds.isEmpty()) {
+            Query.Filter f1 = new Query.FilterPredicate("surveyedLocaleId", Query.FilterOperator.IN, surveyedLocaleIds);
+            Query q = new Query("SurveyInstance");
```

Comment:
I think here you can also save a little by setting it to `setKeysOnly()` so you do not retrieve any other properties. I see you log some form instance properties but unless you think its absolutely necessary, I would suggest we remove them.